src/Http/Routing/src/Internal/MapActionExpressionTreeBuilder.cs

```
@@ -0,0 +1,442 @@
+                var invoker = lambda.Compile();
+                object? defaultBodyValue = null;
+
+                if (allowEmptyBody && bodyType!.IsValueType)
+                {
+                    defaultBodyValue = Activator.CreateInstance(bodyType);
+                }
+
+                requestDelegate = async (target, httpContext) =>
+                {
+                    object? bodyValue;
+
+                    if (allowEmptyBody && httpContext.Request.ContentLength == 0)
+                    {
+                        bodyValue = defaultBodyValue;
+                    }
+                    else
+                    {
+                        try
+                        {
+                            bodyValue = await httpContext.Request.ReadFromJsonAsync(bodyType!);
+                        }
+                        catch (IOException ex)
+                        {
+                            LogRequestBodyIOException(httpContext, ex);
+                            httpContext.Abort();
+                            return;
+                        }
+                    }
+
+                    await invoker(target, httpContext, bodyValue);
+                };
+            }
+            else if (consumeBodyAsForm)
+            {
+                var lambda = Expression.Lambda<Func<object?, HttpContext, Task>>(body, TargetArg, HttpContextParameter);
+                var invoker = lambda.Compile();
+
+                requestDelegate = async (target, httpContext) =>
+                {
+                    // Generating async code would just be insane so if the method needs the form populate it here
+                    // so the within the method it's cached
+                    try
+                    {
+                        await httpContext.Request.ReadFormAsync();
+                    }
+                    catch (IOException ex)
```

Also catch InvalidDataException for bad forms. That one should result in a 400 rather than an abort. cc: @pranavkm

---

Should we log InvalidDataExceptions at the debug level? I suspect you'd see those only when there's an actual bug in the client rather than a transient network issue. I agree it shouldn't result in an error log like an uncaught exception from app code though. How about info level?

---

Yes it should log when it suppresses the exception. I think Debug is the right level if it's going to include the exception, which it should. You wouldn't want exceptions in the info logs.